IllegalAccessErrors trying to run JU5 tests with packagenames that include capital letters.

Thanks to @jbushdiecker for helping to track this down.

Note that this issue appears starting with Micronaut 2.3.0 and still exists in 2.5.0. It worked fine with 2.2.3. This is a blocker for migrating our project beyond version 2.2.3

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create a class in a package that has a capital letter in the name, for example, com.example.A
2. Create  tests for that class
3. Run the tests

### Expected Behaviour
The tests should pass

### Actual Behaviour
The tests don't even run. Micronaut fails to bootstrap. As a result, even a test in a packagename with all lower case fails unless test classes in the packages with caps are removed.

```
Unexpected error loading bean definition [com.example.$A.PackagenamebugATestDefinition]: tried to access class com.example.A.PackagenamebugATest from class com.example.$A.PackagenamebugATestDefinitionClass
io.micronaut.context.exceptions.BeanContextException: Unexpected error loading bean definition [com.example.$A.PackagenamebugATestDefinition]: tried to access class com.example.A.PackagenamebugATest from class com.example.$A.PackagenamebugATestDefinitionClass
	at io.micronaut.context.AbstractBeanDefinitionReference.isPresent(AbstractBeanDefinitionReference.java:92)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:747)
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:721)
	at java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.execLocalTasks(ForkJoinPool.java:1040)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1058)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
Caused by: java.lang.IllegalAccessError: tried to access class com.example.A.PackagenamebugATest from class com.example.$A.PackagenamebugATestDefinitionClass
	at com.example.$A.PackagenamebugATestDefinitionClass.getBeanType(Unknown Source)
	at io.micronaut.context.AbstractBeanDefinitionReference.isPresent(AbstractBeanDefinitionReference.java:84)
	... 15 more
```
### Environment Information

- **Operating System**: Windows 10 and Mac OS 11.3
- **Micronaut Version:** 2.3+ (tested with 2.3.x, 2.4.x and 2.5.x)
- **JDK Version:** AdoptOpenJDK 8.0.232.09 (Windows 10) and 8.0.282 (Mac OS)

### Example Application

- https://github.com/wetted/packagenamebug

--- a/inject-groovy-test/src/main/groovy/io/micronaut/ast/transform/test/AbstractBeanDefinitionSpec.groovy
+++ b/inject-groovy-test/src/main/groovy/io/micronaut/ast/transform/test/AbstractBeanDefinitionSpec.groovy
@@ -89,6 +89,20 @@ abstract class AbstractBeanDefinitionSpec extends Specification {
         }
     }
 
+    @CompileStatic
+    BeanDefinition buildBeanDefinition(String packageName, String className, String classStr) {
+        def beanDefName= '$' + className + 'Definition'
+        String beanFullName = "${packageName}.${beanDefName}"
+
+        def classLoader = new InMemoryByteCodeGroovyClassLoader()
+        classLoader.parseClass(classStr)
+        try {
+            return (BeanDefinition) classLoader.loadClass(beanFullName).newInstance()
+        } catch (ClassNotFoundException e) {
+            return null
+        }
+    }
+
     /**
      * Builds the bean definition for an AOP proxy bean.
      * @param className The class name

--- a/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/visitor/GroovyClassElement.java
+++ b/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/visitor/GroovyClassElement.java
@@ -732,6 +732,16 @@ public class GroovyClassElement extends AbstractGroovyElement implements Arrayab
     }
 
     @Override
+    public String getSimpleName() {
+        return classNode.getNameWithoutPackage();
+    }
+
+    @Override
+    public String getPackageName() {
+        return classNode.getPackageName();
+    }
+
+    @Override
     public boolean isAbstract() {
         return classNode.isAbstract();
     }

--- a/inject-groovy/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
+++ b/inject-groovy/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
@@ -52,4 +52,19 @@ class OuterBean {
 
         definition.intValue(Order).getAsInt() == 10
     }
+
+    void "test a bean definition in a package with uppercase letters"() {
+        when:
+        def definition = buildBeanDefinition('test.A', 'TestBean', '''
+package test.A;
+
+@javax.inject.Singleton
+class TestBean {
+
+}
+''')
+        then:
+        noExceptionThrown()
+        definition != null
+    }
 }

--- a/inject-java-test/src/main/groovy/io/micronaut/annotation/processing/test/AbstractTypeElementSpec.groovy
+++ b/inject-java-test/src/main/groovy/io/micronaut/annotation/processing/test/AbstractTypeElementSpec.groovy
@@ -308,6 +308,18 @@ class Test {
         }
     }
 
+    protected BeanDefinition buildBeanDefinition(String packageName, String className, String cls) {
+        def beanDefName= '$' + className + 'Definition'
+        String beanFullName = "${packageName}.${beanDefName}"
+
+        ClassLoader classLoader = buildClassLoader(className, cls)
+        try {
+            return (BeanDefinition)classLoader.loadClass(beanFullName).newInstance()
+        } catch (ClassNotFoundException e) {
+            return null
+        }
+    }
+
     /**
      * Builds the bean definition for an AOP proxy bean.
      * @param className The class name

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/JavaClassElement.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/JavaClassElement.java
@@ -55,6 +55,9 @@ public class JavaClassElement extends AbstractJavaElement implements ArrayableCl
     private List<PropertyElement> beanProperties;
     private Map<String, Map<String, TypeMirror>> genericTypeInfo;
     private List<? extends Element> enclosedElements;
+    private String simpleName;
+    private String name;
+    private String packageName;
 
     /**
      * @param classElement       The {@link TypeElement}
@@ -512,7 +515,7 @@ public class JavaClassElement extends AbstractJavaElement implements ArrayableCl
         boolean hasNamePredicates = !namePredicates.isEmpty();
         boolean hasModifierPredicates = !modifierPredicates.isEmpty();
         boolean hasAnnotationPredicates = !annotationPredicates.isEmpty();
-        
+
         elementLoop:
         for (Element enclosedElement : enclosedElements) {
             ElementKind enclosedElementKind = enclosedElement.getKind();
@@ -648,8 +651,28 @@ public class JavaClassElement extends AbstractJavaElement implements ArrayableCl
     }
 
     @Override
+    public String getSimpleName() {
+        if (simpleName == null) {
+            simpleName = JavaModelUtils.getClassNameWithoutPackage(classElement);
+        }
+        return simpleName;
+    }
+
+
+    @Override
     public String getName() {
-        return JavaModelUtils.getClassName(classElement);
+        if (name == null) {
+            name = JavaModelUtils.getClassName(classElement);
+        }
+        return name;
+    }
+
+    @Override
+    public String getPackageName() {
+        if (packageName == null) {
+            packageName = JavaModelUtils.getPackageName(classElement);
+        }
+        return packageName;
     }
 
     @Override

--- a/inject-java/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
@@ -133,4 +133,19 @@ public class NumberThingManager extends AbstractThingManager<NumberThing<?>> {}
         definition != null
         definition.getTypeArguments("test.AbstractThingManager")[0].getTypeVariables().get("T").getType() == Object.class
     }
+
+    void "test a bean definition in a package with uppercase letters"() {
+        when:
+        def definition = buildBeanDefinition('test.A', 'TestBean', '''
+package test.A;
+
+@javax.inject.Singleton
+class TestBean {
+
+}
+''')
+        then:
+        noExceptionThrown()
+        definition != null
+    }
 }

--- a/inject/src/main/java/io/micronaut/inject/processing/JavaModelUtils.java
+++ b/inject/src/main/java/io/micronaut/inject/processing/JavaModelUtils.java
@@ -16,6 +16,7 @@
 package io.micronaut.inject.processing;
 
 import io.micronaut.core.reflect.ReflectionUtils;
+import io.micronaut.core.util.StringUtils;
 import io.micronaut.inject.ast.ClassElement;
 import io.micronaut.inject.ast.TypedElement;
 import org.objectweb.asm.Type;
@@ -179,6 +180,53 @@ public class JavaModelUtils {
             }
         } catch (RuntimeException e) {
             return qualifiedName.toString();
+        }
+    }
+
+
+    /**
+     * Get the class name for the given type element without the package. Handles {@link NestingKind}.
+     *
+     * @param typeElement The type element
+     * @return The class name
+     */
+    public static String getClassNameWithoutPackage(TypeElement typeElement) {
+        NestingKind nestingKind;
+        try {
+            nestingKind = typeElement.getNestingKind();
+            if (nestingKind == NestingKind.MEMBER) {
+                TypeElement enclosingElement = typeElement;
+                StringBuilder builder = new StringBuilder();
+                while (nestingKind == NestingKind.MEMBER) {
+                    builder.insert(0, '$').insert(1, enclosingElement.getSimpleName());
+                    Element enclosing = enclosingElement.getEnclosingElement();
+
+                    if (enclosing instanceof TypeElement) {
+                        enclosingElement = (TypeElement) enclosing;
+                        nestingKind = enclosingElement.getNestingKind();
+                    } else {
+                        break;
+                    }
+                }
+                Name enclosingName = enclosingElement.getSimpleName();
+                return enclosingName.toString() + builder;
+            } else {
+                return typeElement.getSimpleName().toString();
+            }
+        } catch (RuntimeException e) {
+            return typeElement.getSimpleName().toString();
+        }
+    }
+
+    public static String getPackageName(TypeElement typeElement) {
+        Element enclosingElement = typeElement.getEnclosingElement();
+        while (enclosingElement != null && enclosingElement.getKind() != ElementKind.PACKAGE) {
+            enclosingElement = enclosingElement.getEnclosingElement();
+        }
+        if (enclosingElement == null) {
+            return StringUtils.EMPTY_STRING;
+        } else {
+            return enclosingElement.toString();
         }
     }
 

